Feature/products

FAEZ10 opened the pull request:
# Pull Request: Add CRUD Routes for Products Module

## **Type of Change**

- [x] Feature
- [ ] Fix

---

## **Description**

This PR introduces the implementation of fully functional CRUD routes for the **Products** module. Key changes include:

- Added API endpoints for creating, retrieving, updating, and deleting products.
- Integrated validation for product creation and update requests using DTOs.
- Standardized API responses using the `ApiResponse` utility class.
- Enhanced error handling for cases like missing or invalid product IDs.
- Documented all routes with Swagger annotations to support API documentation generation.

Relevant motivation and context:

- This feature is essential for managing the **Products** in the application, enabling seamless interaction between the frontend and backend.
- Supports scalability with clean architecture and reusable components.

---

## **How Has This Been Tested?**

This feature has been tested locally using **Postman** and **Swagger**. Tests include:

- [x] Creating products with valid and invalid payloads.
- [x] Retrieving all products and specific products by ID.
- [x] Updating existing products and handling non-existent IDs.
- [x] Deleting products and validating proper status codes for each scenario.

---

## **Checklist**

- [x] My code follows the style guidelines of this project.
- [x] I have performed a self-review of my own code.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have made corresponding changes to the documentation.
- [x] My changes generate no new warnings.
- [x] I have added tests that prove my feature works.
- [x] New and existing unit tests pass locally with my changes.
- [x] Any dependent changes have been merged and published in downstream modules.

FAEZ14 committed:
Feat(routes): add CRUD routes for Products module

- Implemented routes for creating, retrieving, updating, and deleting products.
- Integrated validation middleware for product creation and updates.
- Added error handling for non-existent products in get, update, and delete operations.
- Standardized API responses using ApiResponse utility.
- Documented routes with Swagger annotations for API documentation.

FAEZ14 committed:
refactor(service): align ProductService methods with MongooseRepository

- Renamed methods in ProductService to match the conventions of MongooseRepository:
  - findAll -> getAll
  - findById -> getById
  - update -> updateById
  - delete -> deleteById
- Improved consistency and maintainability by adhering to standardized method names.
- Ensured all references to ProductRepository methods are updated accordingly.

DaniOmer commented on src/apps/product/domain/product.dto.ts:
```diff
@@ -0,0 +1,13 @@
+  @IsNotEmpty()
+  @MinLength(3)
+  @MaxLength(255)
+  readonly name!: string;
```
Il faudrait valider que c'est une string

DaniOmer commented on src/apps/product/domain/product.dto.ts:
```diff
@@ -0,0 +1,13 @@
+  @IsNotEmpty()
+  @MinLength(10)
+  @MaxLength(1000)
+  readonly description?: string;
```
Validation de string aussi. le plus pertinent entre IsAlpha et IsAlphaNumeric

DaniOmer commented on src/apps/product/domain/product.service.ts:
```diff
@@ -0,0 +1,38 @@
+    const product = await this.productRepository.getById(productId); 
+    if (!product) {
+      throw new Error("Product not found");
+    }
```
throw BadRequestError pour que le middleware le catch proprement

DaniOmer commented on src/apps/product/domain/product.service.ts:
```diff
@@ -0,0 +1,38 @@
+    const product = await this.productRepository.updateById(productId, updates); 
+    if (!product) {
+      throw new Error("Product not found");
+    }
```
throw BadRequestError pour que le middleware le catch proprement

DaniOmer commented on src/apps/product/api/product.route.ts:
```diff
@@ -0,0 +1,157 @@
+import { validateDtoMiddleware } from "../../../librairies/middlewares/validation.middleware";
+import { ProductCreateDTO } from "../domain/product.dto";
+
+const router = Router();
```
ici on peut éviter de créer une nouvelle instance de router. pour cela on peut avoir une fonction qui contient les routes et return router `export default (router: Router) => { 
} `

FAEZ14 committed:
refactor(service & routes): enhance error handling, DTO validation, and router configuration

- **ProductService**:
  - Replaced generic errors with structured `BadRequestError` for better error handling.
  - Added detailed error context (`productId`) in methods like `getProductById` and `updateProduct`.
  - Improved consistency of error responses with `message`, `code`, and `context`.
  - Renamed `listProducts` to `getAllProducts` for better naming consistency.

- **ProductCreateDTO**:
  - Added `@IsString` validation to ensure fields are explicitly validated as strings.
  - Combined `@IsNotEmpty`, `@MinLength`, and `@MaxLength` validations for robust input checks.

- **Product Routes**:
  - Refactored routes to avoid direct instantiation of `Router`.
  - Implemented a function-based router configuration to inject the `Router` instance dynamically.
  - Preserved Swagger annotations for all routes.

FAEZ14 committed:
Merge branch 'dev' into feature/products

FAEZ14 committed:
feat: Implement generic BaseService and validateIdMiddleware for cleaner and reusable code

- Added a `BaseService` class to centralize CRUD operations for all services, reducing redundancy.
- Created `validateIdMiddleware` to validate MongoDB ObjectId in route parameters, ensuring clean and reusable middleware logic.
- Updated `ProductService` and `ProductRepository` to leverage `BaseService` for shared functionality.
- Enhanced code maintainability and scalability by promoting a generic and modular structure.

FAEZ14 committed:
feat: Improve error handling for missing entity IDs in BaseService

- Enhanced `BaseService` to handle cases where an entity ID is not found.
- Added `validateEntityExists` method to centralize and standardize "not found" error handling.
- Included detailed context in error messages, such as the operation type, entity name, and timestamp.
- Ensured that all "not found" scenarios

DaniOmer commented on src/librairies/services/base.service.ts:
```diff
@@ -0,0 +1,63 @@
+    getById: (id: string) => Promise<T | null>;
+    getAll: () => Promise<T[]>;
+    updateById: (id: string, updates: Partial<T>) => Promise<T | null>;
+    deleteById: (id: string) => Promise<T | null>;
```
Je pense que la classe BaseService n'est pas très pertinente dans le context du projet. la plupart des applications devront la surchargée ou réécrire des méthodes comme create ou updateById. Par exemple la création de produit peut nécessiter des vérifications de stock, de catégories ; pareil pour payment qui fera certainement d'autres appels ou calcul avant de créer/update

FAEZ10 replied:
Mais après pour les models basiques comme color, size, ... ça facilite et surtout ça permet d'avoir une nomenclature  commune, c'est pas un problème si t'as des spécificités  de surcharger les méthodes.

FAEZ10 replied:
Au pire si t'as pas besoin tu l'extends pas

DaniOmer replied:
Oui mais si t'as un BaseService c'est pour que tous tes services l'utilisent sinon on n'utilise pas bien la POO dans ce cas

FAEZ10 replied:
Une base ça reste une base elle est pas censée gérer les spécificité aussi c'est pour cela qui a la possibilité de surcharger

FAEZ10 replied:
En plus y'a que les méthodes basiques

DaniOmer replied:
je comprends l'idée mais les services comme (Order, Payment, Invoice, etc.) ont des logiques trop différentes, et surcharger chaque méthode va vite complexifier le code, ça risque d'être moins lisible et plus difficile à maintenir. Si l'idée c'est d'avoir une même nomenclature autant utiliser une interface pour éviter le couplage, sinon on pourrait alors faire la même chose pour les controller et rajouter dans méthodes dans `BaseController`

FAEZ10 replied:
Ce que je prends pas c'est que les méthodes create, getAll,  le getById, updateById et deleteById,  ces des méthodes qui vont pas trés bouger car même si t'as une logique un peu différente tu faire ta logique et les appeler après avec le super car c'est spécifique à chaque model

FAEZ10 replied:
Elles sont assez génériques

DaniOmer replied:
Encore une fois je comprends l'idée, mais il faut rappeler le principe d'une classe de base : elle est censée fournir des fonctionnalités génériques réutilisables sans nécessiter de modifications majeures dans les classes dérivées. Si chaque service doit surcharger des méthodes ou ajouter une logique spécifique avant d'appeler super, ça va à l'encontre de cet objectif, car on complexifie au lieu de simplifier.

DaniOmer replied:
et dans notre projet actuel, quasiment toutes les apps vont surchargés deux à trois méthodes. ça complexie grandement le code et la maintenabilité. ma question c'est pourquoi tu n'as pas fait de même pour les controller ?

FAEZ10 replied:
Pour répondre à ta question c'est effectivement dans le controller que tu vas gérer la logique métier de création le service tu vas l'appeler juste au moment de créer, si y'a des checks à faire ou des dépendance d'autres table tu les fais une fois t'as toute ta data tu appelles le service.create. En tout cas c'est comme ça que moi je vois les choses

DaniOmer replied:
okay je comprends l'idée, mais je pense qu'il est important de garder en tête que le pattern service est utilisé pour séparer la logique métier de la gestion des requêtes HTTP (ce que fait le contrôleur). Si on fait ce que tu propose, il n'y aura plus de séparation des responsabilités

DaniOmer replied:
et  le repository et le service feront la même chose finalement

FAEZ10 replied:
Par contre je suis d'accord sur le fait que le service et le repository vont faire presque la même chose.

FAEZ10 replied:
Tu suggères quoi du coup monsieur le CTO ?

DaniOmer replied:
chaque service doit gérer ces logiques métier et au vu des divergences que l'on peut avoir pas pertinent d'avoir cette classe BaseService qui fait la même chose que BaseRepository finalement. à la limite la BaseService peut contenir la méthode de validation qui renvoie une erreur si pas de data, trop de refacto va introduire de la complexité pas nécessaire

johnnyhelloworld replied:
Le service utilise le repository pour accéder aux données tout en appliquant des règles métier

DaniOmer commented on src/librairies/services/base.service.ts:
```diff
@@ -0,0 +1,63 @@
+   */
+  private validateEntityExists(entity: T | null, id: string, operation: string): T {
+    if (!entity) {
+      throw new BadRequestError({
```
Cette méthode ne valide pas l'existence de l'entité mais plutôt la data retourné par  le repository. ça peut prêter à confusion

FAEZ10 replied:
En fait y'a une différence subtile avec le isValidObjectId il valid si l'id est correct en terme de format mais si par exemple l'id est comme ça 674375b8822fff6806be7557 et que tu change le 7 en mettant 6 c'est un bon format du coup il passe le middlware et après il va rien trouvé car l'id n'existe pas du coup t'aura une réponse success avec 0 data

FAEZ10 replied:
Alors que t'es pas censé faire des appels avec des id qui n'existe pas sauf si tu veux hacker

DaniOmer replied:
ouais je comprends et c'est logique qu'il retourne null si l'id est incorrect. c'est ce que fait `findByIdAndUpdate`. mais ta méthod vérifie plutôt la data qui est retournée et non pas l'entité donc le nommer peut-être différemment

FAEZ10 replied:
Oui j'avoue

FAEZ10 replied:
Je vais la renommer comme ça : validateDataExists

DaniOmer commented on src/librairies/services/base.service.ts:
```diff
@@ -0,0 +1,63 @@
+          entity: this.entityName,
+          operation,
+          timestamp: new Date().toISOString(),
+        },
```
le message d'erreur et context c'est pour nous développeur. Je pense un message comme `Product not found est largement suffisant` pas besoin de spécifier l'id, pareil dans le context

FAEZ10 replied:
J'ai supprimer le context j'ai gardé `${this.entityName} not found for ID: ${id}`, j'ai laissé l'id que ça donne l'information comme quoi c'est à cause de l'id que la route est not found

DaniOmer replied:
ok si tu veux mais c'est trop d'infos pour pas grande chose. Juste `product not found`, on sait que c'est lié à l'id

FAEZ14 committed:
feat: refactor ProductService and BaseService for better separation of concerns

- Moved generic `validateDataExists` logic to `BaseService` for reuse across different services.
- Simplified `BaseService` to only include reusable methods for entity validation.
- Shifted CRUD operations (`create`, `getAll`, `getById`, `updateById`, `deleteById`) back to `ProductService` to handle specific business logic.
- Enhanced flexibility and maintainability by allowing each service to define its specific logic while leveraging shared validation methods from `BaseService`.
- Prepared the structure for easier extension to other services like `ColorService` and `SizeService`.

This change improves code organization, reusability, and scalability while reducing duplication of entity validation logic.

DaniOmer commented on src/librairies/middlewares/validateId.middleware.ts:
```diff
@@ -0,0 +1,33 @@
+            entity: entityName,
+            operation: "validateId",
+            timestamp: new Date().toISOString(),
+          },
```
le timestamp n'est pas nécessaire. le logger utilisé par le `errorHandlerMiddleware` nous permet déjà d'avoir le timestamp. et je pense qu'on ne devrait pas donner des message d'erreur trop explicites comme `Invalid ${entityName} ID format`.  ça donnes des piste à un attaquant qui cherche à savoir quels format d'id est correct pour faire du brute force par exemple même si ça parait compliqué. `Bad request` ou à la limite `Product not found` serait plus secure. si tu veux qu'on ait plus de détails, peut-être utiliser le logger

FAEZ14 committed:
refactor(core): enhance BaseService and middleware for cleaner and secure error handling

- Improved `BaseService` with `validateDataExists` method to allow overriding the default entity name, enabling more flexible and reusable error messages.
- Simplified error message structure in `validateDataExists` for clarity and security.
- Refactored `validateIdMiddleware` to provide concise error messages, minimizing information exposure and enhancing security.

FAEZ14 committed:
refactor(routes): use Router instance directly in product routes for clarity and consistency
